src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -671,18 +671,10 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
                 ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
             }
 
-            // kept outside of the small arrays hot path to have inlining without big size growth
-            return AllocateNewUninitializedArray(length, pinned);
-
-            // remove the local function when https://github.com/dotnet/runtime/issues/5973 is implemented
-            static T[] AllocateNewUninitializedArray(int length, bool pinned)
-            {
-                GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
-                if (pinned)
-                    flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
-
-                return Unsafe.As<T[]>(AllocateNewArray(typeof(T[]).TypeHandle.Value, length, flags));
-            }
+            GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
```

Comment:
The optimization does work kick in for shared generic code (when the caller is generic and T comes via dictionary lookup). Is this going to be a regression for that case?

Reply:
It seems that this internal method was always inlined anyway so I only cleaned it up. I'll check the shared T case

Reply:
I've just checked - this method is inlined even for shared context, thus, it should be always inlined anyway (unless hits some inliner's budget limitations):
```asm
; Assembly listing for method System.GC:AllocateUninitializedArray[System.__Canon](int,bool):System.__Canon[]
G_M41526_IG01:  ;; offset=0000H
       push     rsi
       sub      rsp, 48
       mov      qword ptr [rsp+28H], rcx
       mov      esi, edx
G_M41526_IG02:  ;; offset=000CH
       test     r8b, r8b
       jne      SHORT G_M41526_IG08
       mov      rdx, qword ptr [rcx+38H]
       cmp      qword ptr [rdx+08H], 32
       jle      SHORT G_M41526_IG05
G_M41526_IG03:  ;; offset=001CH
       mov      rax, qword ptr [rdx+20H]
       test     rax, rax
       je       SHORT G_M41526_IG05
G_M41526_IG04:  ;; offset=0025H
       mov      rcx, rax
       jmp      SHORT G_M41526_IG06
G_M41526_IG05:  ;; offset=002AH
       mov      rdx, 0x7FF7A35DC930      ; global ptr
       call     CORINFO_HELP_RUNTIMEHANDLE_METHOD
       mov      rcx, rax
G_M41526_IG06:  ;; offset=003CH
       movsxd   rdx, esi
       call     CORINFO_HELP_NEWARR_1_OBJ
       nop      
G_M41526_IG07:  ;; offset=0045H
       add      rsp, 48
       pop      rsi
       ret      
G_M41526_IG08:  ;; offset=004BH
       mov      rcx, qword ptr [rcx+38H]
       mov      rcx, qword ptr [rcx]
       call     CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
       mov      rcx, rax
       call     [System.ThrowHelper:ThrowInvalidTypeWithPointersNotSupported(System.Type)]
       int3   
```

Reply:
Ah, it is not inlined but is always eliminated actually. `AllocateUninitializedArray` always uses `new T[]` for shared T (or throws an exception if pinned is true)

Reply:
> AllocateUninitializedArray always uses new T[] for shared T

It uses newT[] for T with generic pointers. You can have T that is shared and does not contain generic pointers. For example, try this on `MyStruct<string>` with `struct MyStruct<T> { int value; }`.